taca_ngi_pipeline/deliver/deliver_castor.py

```diff
@@ -0,0 +1,170 @@
+    
+
+    def deliver_project(self):
+        """ Deliver all samples in a project to castor
+            
+            :returns: True if all samples were delivered successfully, False if
+                any sample was not properly delivered or ready to be delivered
+        """
+        try:
+            logger.info("Delivering {} to {}".format(
+                str(self), self.expand_path(self.castordeliverypath)))
+            if self.get_delivery_status() == 'DELIVERED' \
+                    and not self.force:
+                logger.info("{} has already been delivered".format(str(self)))
+                return True
+            # right now, don't catch any errors since we're assuming any thrown
+            # errors needs to be handled by manual intervention
+            status = True
+            # Open sftp session with castor, in this way multiple tranfer will be possible
+            try:
+                transport=paramiko.Transport(self.castorsftpserver)
+                password = getpass.getpass(prompt='Bianca/Castor Password for user {}:'.format(self.castorsftpserver_user))
+                transport.connect(username = "{}-{}".format(self.castorsftpserver_user, self.uppnexid), password = password)
+            except Exception as e:
+                logger.error("Caught exception: {}: {}".format(e.__class__, e))
+                raise
+
+            #memorize all samples that needs to be delivered
+            samples_to_deliver = [sentry['sampleid'] for sentry in db.project_sample_entries(
+                    db.dbcon(), self.projectid).get('samples', [])]
+            status = True
+            #open the sftp client
+            #sftp_client = transport.open_sftp_client()
+            sftp_client = MySFTPClient.from_transport(transport)
+            # move to the delivery directory in the sftp
+            #open one client session and leave it open for all the time of the transfer
+            sftp_client.chdir(self.expand_path(self.castordeliverypath))
+            #create the project folder in the remote server
+            if not rexists(sftp_client, self.projectid):
+                sftp_client.mkdir(self.projectid)
+            #move inside the project folder
+            sftp_client.chdir(self.projectid)
+            #now cycle across the samples
+            for sampleid in samples_to_deliver:
+                sampleDelivererObj = CastorSampleDeliverer(self.projectid, sampleid, sftp_client)
+                st = sampleDelivererObj.deliver_sample()
+                status = (status and st)
+            try:
+                transport.close()
+            except Exception as e:
+                logger.error("Caught exception: {}: {}".format(e.__class__, e))
+                raise
+            # close the client
+            self.sftp_client.close()
+            # query the database whether all samples in the project have been sucessfully delivered
+            if self.all_samples_delivered():
```

Comment:
should this be done before closing the connections so that if closing throws exceptions, the delivery status would still be updated? Or could it be that if you get an exception while closing connections, files might not have been delivered properly? In that case, I think this order makes sense.